### PR TITLE
Partial web platform support

### PR DESCRIPTION
**Pull request by oxyii**
**Good news! Still working on native platforms via expo! (Tested on Android 8.1 and iOS 12.1.4)**

1. Rewrited scripts in `package.json`. Use `expo start` for native platforms (`expo-cli` must be installed) and `npm start` for Web platform.

2. Fixed 3 icons on Settings screen. I don't know why but the same icons have a separate names. e.g. see `ios-plane` [here](https://infinitered.github.io/ionicons-version-3-search/) and `ios-airplane` [here](https://expo.github.io/vector-icons/).

3. I could not start MaterialCommunityIcons on web (((

4. I'm surprised that you use icons from `react-native-vector-icons` instead RNE Icon ))) e.g. [here](https://github.com/Monte9/react-native-elements-app/blob/master/src/views/login/screen1.js#L12). Fixed.

5. Updated SDK to v32. I'm confused that in app using expo v32 but RN SDK v31.

**Bad news now**

1. Non usable navigation for web. Address place in browser are not changing by screen changes.

2. UI. It takes a lot of work (((

I will return to work on this repo later

### CODE REVIEW

**Comment by Monte9** [comment]
👏 awesome work so far @oxyii! really appreciate you taking the lead on this.

I'll take a look at this in a couple days and get back to you.

@xavier-villelegier or @iRoachie if you guys have any thoughts please feel free to jump in here.

**Comment by iRoachie** [comment]
Man this is a huge undertaking, awesome job 💯 In your testing did you see anything that doesn't work as expected?

**Comment by oxyii** [comment]
@iRoachie Yep. `ListItem` is not working. Fixed by [this commit](https://github.com/Monte9/react-native-elements-app/pull/74/commits/75407e0d8a09d275b73e067e63529cff329c621e)

Just change [this](https://github.com/react-native-training/react-native-elements/blob/master/src/list/ListItem.js#L351) with [this](https://github.com/Monte9/react-native-elements-app/pull/74/commits/75407e0d8a09d275b73e067e63529cff329c621e#diff-787867683102e8fe39dcdad9240764a0R354) (React.forwardRef) and it's working!

And `linearGradientProps` prop on `ListItem` requires ViewComponent implement, something like [this](https://github.com/expo/expo/tree/master/packages/expo-linear-gradient/src)

**Comment by oxyii** [comment]
`font-family` looks correct because it works normally on native platforms. I still don't understand why on the web there are problems with it

**Comment by oxyii** [comment]
assets folder can be placed on root path of project (by default for expo). Thanks to [removeModuleScopePlugin func](https://github.com/arackaf/customize-cra#removemodulescopeplugin).

**Comment by oxyii** [comment]
@iRoachie I understood `ListItem` problems. My idea with `React.forwardRef` will not solve it.

[PadView](https://github.com/react-native-training/react-native-elements/blob/master/src/list/ListItem.js#L351) should not be a stateless function. He must be a class with `setNativeProps` implementation.

```jsx
import React from "react";
import { View } from 'react-native';

class PadView extends React.Component {
  setNativeProps(props) {
    this.refs['PAD_CONTAINER_REF'].setNativeProps(props);
  }

  render() {
    const {children, pad, Component, ...props} = this.props;
    const childrens = React.Children.toArray(children);
    const { length } = childrens;
    const Container = Component || View;
    return (
      <Container ref={'PAD_CONTAINER_REF'} {...props}>
        {React.Children.map(
          childrens,
          (child, index) =>
            child && [child, index !== length - 1 && <View width={pad} />]
        )}
      </Container>
    );
  }
}
```

`ListItem` can remain as it is but for web we need to set `ViewComponent` e.g. from `expo-linear-gradient`.

**Comment by oxyii** [comment]
@Monte9 @iRoachie  I return to the main work. I will continue working with this repo few days later.

Finally: ListItem works fine with `React.forwardRef PadView` (see [RN issue](https://github.com/facebook/react-native/issues/1040)) but web-platform developers **must** set `ViewComponent` for using gradient on `ListItem`, `Button` or other components. My implementation of `ViewComponent` in this app see [here](https://github.com/oxyii/react-native-elements-app/tree/master/src/components/LinearGradient). `expo-linear-gradient` cannot be a dep for RNE because installation process requires changes into platforms sources but `expo-linear-gradient` can be dep for app.

Also I see paddings bugs on `ListItem`...

And as I say 

> Non usable navigation for web

> UI. It takes a lot of work (((

**Comment by oxyii** [comment]
I hope PR ready for merging now )))

gh-pages (with header on all screens): https://oxyii.github.io/react-native-elements-app

**Comment by Monte9** [comment]
@oxyii it looks great. Can you please rebase as I merged in a cleanup PR by @iRoachie!

Also let me know when it's ready for review.

**Comment by oxyii** [comment]
@Monte9 All conflicts are resolved, gh-pages updated too

**Comment by Monte9** [comment]
@oxyii ready to review & merge PR?

**Comment by oxyii** [comment]
@Monte9 Yep )

**Comment by oxyii** [comment]
@Monte9 I have not added a hamburger-menu to PR. It is required on all screens to navigate on web platform, but this menu is incompatible with the design of some screens, such a profile screen. Please decide whether to include this menu into PR or not. Diff for adding hamburger-menu you can see [here](https://github.com/oxyii/react-native-elements-app/compare/master...oxyii:all-pages-with-headers#diff-8df0439da9cfdc5038840bd1fb1f8669).
Выделение_001

**Comment by Monte9** [comment]
@oxyii good addition. Thanks for pointing it out. I think even though it doesn't directly fit in with the design, it's needed for easy & consistent navigation across all screens.  ✅

File: package.json [hunk 1]
```diff
@@ -4,28 +4,48 @@
...
+    "react-dom": "^16.8.3",
+    "react-native": "https://github.com/expo/react-native/archive/sdk-32.0.0.tar.gz",
     "react-native-elements": "^1.1.0",
-    "react-native-ratings": "^5.3.0",
+    "react-native-ratings": "^6.3.0",
     "react-native-touchable-scale": "2.0.0",
-    "react-navigation": "1.5.8"
+    "react-native-vector-icons": "^6.3.0",
+    "react-native-web": "^0.10.0",
+    "react-navigation": "1.5.8",
```
Comment (Monte9) [inline package.json]: We can make this part of another PR, but we should update this to v3

File: package.json [hunk 2]
```diff
@@ -4,28 +4,48 @@
...
     "clean-install": "rm -rf node_modules && npm cache clean --force && watchman watch-del-all && yarn"
   },
   "devDependencies": {
+    "@babel/plugin-proposal-class-properties": "^7.3.4",
     "babel-preset-expo": "^5.0.0",
-    "prettier": "^1.15.3"
-  }
+    "customize-cra": "^0.2.12",
+    "prettier": "^1.15.3",
+    "react-app-rewired": "^2.1.0"
```
Comment (Monte9) [inline package.json:45]: what do you use `customize-cra` & `react-app-rewired` for? never heard/used those before but they sound cool

File: src/components/LinearGradient/index.js [hunk 3]
```diff
@@ -0,0 +1 @@
+export {LinearGradient} from "./LinearGradient"
```
Comment (Monte9) [inline src/components/LinearGradient/index.js:1]: can you explain how this works? So this index file essentially exports either `.js` if it's a mobile paltform or `.web.js` if it's web? Which library handles the logic for this?

File: src/serviceWorker.js [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+        });
+}
+
+export function unregister() {
+    if ('serviceWorker' in navigator) {
+        navigator.serviceWorker.ready.then(registration => {
+            registration.unregister();
+        });
+    }
+}
```
Comment (Monte9) [inline src/serviceWorker.js:135]: did you get this file from CRA?

**Review by Monte9 — APPROVED** [review]
Great progress! 👏 Can you please submit $100 on OpenCollective as a "Thank you" from the RNE Team for putting in all this work. 😄

Here is how to do it:

1. Go on [Open Collective](https://opencollective.com/react-native-elements)
2. Click "Submit Expense"
**Description:** 1.0.0 PRs and issues
**Category:** Engineering

3. For the receipt, use [this template](https://docs.google.com/document/d/1Pc9ic3tbgqqhVUIF-T6QEf8B3d9IVtP_GGnK9LCVcY0/edit?usp=sharing)

**Comment by Monte9** [comment]
Also what screens are compatible/visible in RNW? I only see the `.web.js` for a few files.

Overall nice work. I left a few questions that I am mostly curious about.

**Comment by oxyii** [comment]
@Monte9 I added ham-menu.

**About `*.web.js` files**: RN supports platform specific code. if I have not added a separate `*.web.js`, then the main `*.js` works well on web too. All screens works fine on web, see my [gh-pages](https://oxyii.github.io/react-native-elements-app).

```bash
AnotherComponent.js              <- for all platforms
# BUT
SomeComponent.js                 <- Will never be called on our env but can be called if you add something crazy like windows desktop platform support
SomeComponent.ios.js             <- For iOS
SomeComponent.native.js          <- Will call on android because .android.js is not exists
SomeComponent.web.js             <- For web
```

See [RN Docs](https://facebook.github.io/react-native/docs/platform-specific-code#platform-specific-extensions) and [one more](https://facebook.github.io/react-native/docs/out-of-tree-platforms) and [CRA webpack config](https://github.com/facebook/create-react-app/blob/master/packages/react-scripts/config/webpack.config.js#L277) for details.

File: src/serviceWorker.js (hunk 4 above)
Comment (oxyii) [reply src/serviceWorker.js:135]: Yep, It's standard React service worker.

File: src/components/LinearGradient/index.js (hunk 3 above)
Comment (oxyii) [reply src/components/LinearGradient/index.js:1]: See [comment below](https://github.com/Monte9/react-native-elements-app/pull/74#issuecomment-473881564)

File: package.json (hunk 2 above)
Comment (oxyii) [reply package.json:45]: It's for `eject` escaping. See [my comment](https://github.com/react-native-training/react-native-elements/issues/1808#issuecomment-468869382)

**Comment by oxyii** [comment]
Okay. Real world example:

```bash
Touchable.js                  <- export TouchableOpacity
Touchable.android.js          <- export TouchableNativeFeedback
```

then you can just import `from Touchable` and will be imported `TouchableNativeFeedback` on android and `TouchableOpacity` on all others platform include iOS, web, tvOS, [Qt](https://github.com/status-im/react-native-desktop) and [windows-desktop](https://github.com/Microsoft/react-native-windows).

File: package.json (hunk 1 above)
Comment (oxyii) [reply package.json]: I believe that navigation v3 will work. See [docs](https://reactnavigation.org/docs/en/web-support.html)

**Comment by oxyii** [comment]
@Monte9 that kid (@fungilation) blocked me in issue ))) About docs:

if app started not ejected (via `expo init`) then [this comment](https://github.com/react-native-training/react-native-elements/issues/1808#issuecomment-468869382). Get `config-overrides.js` contents from PR better. But if app started already ejected (via CRNWA) then use [this comment](https://github.com/react-native-training/react-native-elements/issues/1808#issuecomment-469471406)

**Comment by Monte9** [comment]
haha ok sounds good. I'll follow up on the issue & make sure to update the docs! 👍

**Comment by oxyii** [comment]
Yes! `react-navigation` v3 is here!

* Fixed drawer width for desktop (see [here](https://github.com/oxyii/react-native-elements-app/blob/master/src/App.js#L77))
* and, yes, `react-navigation` v3 included!

navigation v3 depends on [react-native-gesture-handler](https://github.com/kmagiera/react-native-gesture-handler) but RNGH web platform support was [published](https://github.com/kmagiera/react-native-gesture-handler/releases/tag/1.1.0) a few days ago. So expo v32 includes old `1.0.17` (`~1.0.14` exactly) version of RNGH that not works on web. A new version of RNGH seems to be included in expo v33 (see [here](https://github.com/expo/expo/blob/master/packages/expo/package.json#L131)). To make it all work on all platforms for now I had to use a little magic: [include both versions](https://github.com/oxyii/react-native-elements-app/blob/master/package.json#L20-L21) and [make alias for it for web only](https://github.com/oxyii/react-native-elements-app/blob/master/config-overrides.js#L42). This is just for now, until expo v33 will be published.

About navigation v3 web support [recommendations](https://reactnavigation.org/docs/en/web-support.html): it's not working for drawer navigation... `react-navigation-drawer` is trying to insert something strange into browser history (replace origin blocked in browser). But without web platform support recommendations all works fine on web too but without browser history changes again.

[gh-pages](https://oxyii.github.io/react-native-elements-app/) updated. See bottom tab bar on `Components` drawer item. He must be white instead blue. Clear app cache if not.

**Comment by Monte9** [comment]
@oxyii thanks for the detailed explanation.

Nice job on the `react-navigation` v3 upgrade! Looks like you had some trouble with the drawer in v3 but managed to figure it out! 👏 

Also I played around with the app deployed on `gh-pages` and I am thoroughly impressed. Mostly all the screens worked nicely and looked half decent. Imagine if we improve the `.web.js` components and optimize it for web. That would be a fantastic demo and will make RNE + RNW easily accessible to most people. 💯 🔥 

Going to merge this PR in now! Thanks again for all your work. I have approved your open collective request. 🤑

**Comment by Monte9** [comment]
Couple other thoughts:
- Once it's merged in, I will try to do a fresh clone & try to get the app & web working locally. Are there any setup steps for web that I should follow?
- How does the `gh-pages` deploy currently? Let's document this process and add it to the readme.
- I love your work on this project I would like you to continue helping us out as much as you can. I think RNE + RNW can be a really big advantage & it can grow to be much bigger than it is right now. I will plan to move this repo & RNE repo to the new org https://github.com/react-native-elements and have invite you to be a collab.

**Comment by oxyii** [comment]
@Monte9 

* See initial message of PR

> Use expo start for native platforms (expo-cli must be installed) and npm start for Web platform.

after install deps )

* Web is a **simple React application** [bundled](https://github.com/Monte9/react-native-elements-app/pull/74/files#diff-b9cfc7f2cdf78a7f4b91a753d10865a2R31) in `build` folder! That's it! So `gh-pages` deployed [simple](https://medium.com/the-andela-way/how-to-deploy-your-react-application-to-github-pages-in-less-than-5-minutes-8c5f665a2d2a) as any other React app.

* I'm in )))

**Comment by oxyii** [comment]
@Monte9 My current cross-platform project have `package.json` scripts like below

```json
  "scripts": {
    "web": "react-app-rewired start",
    "native": "expo start",
    "build:web": "react-app-rewired build",
    "build:android": "expo ba",
    "build:ios": "expo bi",
    "eject:web": "react-scripts eject",
    "eject:native": "expo eject"
  },
```

So `yarn web` (or `npm run web`) will start web-dev server, `yarn native` (or `npm run native`) will start metro bundler etc. It seems convenient

**Review by Monte9 — APPROVED** [review]
✅ 💯 👏

**Comment by Monte9** [comment]
@oxyii would you like to join the core RNE team and help us move RNE foward?!

If yes, please join our slack here: https://reactnativetraining.herokuapp.com
and then DM me @monte9 on slack and I'll invite you to our team. 😄

**Comment by oxyii** [comment]
@Monte9 I'm in
